examples/dreamer/dreamer.py

```diff
@@ -0,0 +1,310 @@
+            if not isinstance(transform, VideoRecorder):
+                recorder_rm.append_transform(transform)
+    else:
+        recorder_rm = recorder
+
+    if isinstance(create_env_fn, ParallelEnv):
+        recorder_rm.load_state_dict(create_env_fn.state_dict()["worker0"])
+        create_env_fn.close()
+    elif isinstance(create_env_fn, EnvCreator):
+        recorder_rm.load_state_dict(create_env_fn().state_dict())
+    else:
+        recorder_rm.load_state_dict(create_env_fn.state_dict())
+    # reset reward scaling
+    for t in recorder.transform:
+        if isinstance(t, RewardScaling):
+            t.scale.fill_(1.0)
+            t.loc.fill_(0.0)
+
+    record = Recorder(
+        record_frames=cfg.record_frames,
+        frame_skip=cfg.frame_skip,
+        policy_exploration=policy,
+        recorder=recorder,
+        record_interval=cfg.record_interval,
+        log_keys=cfg.recorder_log_keys,
+    )
+
+    final_seed = collector.set_seed(cfg.seed)
+    print(f"init seed: {cfg.seed}, final seed: {final_seed}")
+    ## Training loop
+    collected_frames = 0
+    pbar = tqdm.tqdm(total=cfg.total_frames)
+    r0 = None
+    path = Path('./log')
+    path.mkdir(exist_ok=True)
+    
+    scaler= GradScaler()
+    
+    for i, tensordict in enumerate(collector):
+
+        # update weights of the inference policy
+        collector.update_policy_weights_()
+
+        if r0 is None:
+            r0 = tensordict["reward"].mean().item()
+        pbar.update(tensordict.numel())
+
+        current_frames = tensordict.numel()
+        collected_frames += current_frames
+        replay_buffer.extend(tensordict.cpu())
+
+        if collected_frames >= cfg.init_env_steps:
+            for j in range(cfg.optim_steps_per_batch):
+                # sample from replay buffer
+                sampled_tensordict = replay_buffer.sample(cfg.batch_size).to(device)
+
+                with autocast():
+                    model_loss_td, sampled_tensordict = world_model_loss(sampled_tensordict)
+                    actor_loss_td, sampled_tensordict = actor_loss(
+                        sampled_tensordict
+                    )
+                    value_loss_td, sampled_tensordict = value_loss(
+                        sampled_tensordict
+                    )                
+                
+                scaler.scale(model_loss_td["loss_world_model"]).backward()
+                scaler.scale(actor_loss_td["loss_actor"]).backward()
+                scaler.scale(value_loss_td["loss_value"]).backward()
+
+                scaler.unscale_(world_model_opt)
+                clip_grad_norm_(world_model.parameters(), cfg.grad_clip)
+                scaler.unscale_(actor_opt)
+                clip_grad_norm_(actor_model.parameters(), cfg.grad_clip)
+                scaler.unscale_(value_opt)
+                clip_grad_norm_(value_model.parameters(), cfg.grad_clip)
+
+                scaler.step(world_model_opt)
```

Comment:
I think we should do 

loss1 = ...
Optim1.step()

Loss2 = ...
Optim2.Step()
Etc

Like this we allow the gpu to free mem when calling backward

Reply:
I was doing it like that before, however the pb is that autocast struggle in this context. Indeed according to pytorch doc only a single scaler can be created and you cannot scale again after unscale_ . I felt that 16bit precision would be better to have

Reply:
Can't you scale / unscale multiple times? What would be the difference between that and scaling/unscaling through a loop?

Reply:
```python
                with autocast(dtype=torch.float16):
                    model_loss_td, sampled_tensordict = world_model_loss(
                        sampled_tensordict
                    )
                    if cfg.record_video:
                        world_model_td = sampled_tensordict.clone().select(
                            "pixels", "reco_pixels", "posterior_states", "next_belief"

                        )[:4].detach()
                scaler1.scale(model_loss_td["loss_world_model"]).backward()
                scaler1.unscale_(world_model_opt)
                clip_grad_norm_(world_model.parameters(), cfg.grad_clip)
                scaler1.step(world_model_opt)
                world_model_opt.zero_grad()
                scaler1.update()

                with autocast(dtype=torch.float16):
                    actor_loss_td, sampled_tensordict = actor_loss(sampled_tensordict)
                scaler2.scale(actor_loss_td["loss_actor"]).backward()
                scaler2.unscale_(actor_opt)
                clip_grad_norm_(actor_model.parameters(), cfg.grad_clip)
                scaler2.step(actor_opt)
                actor_opt.zero_grad()
                scaler2.update()

                with autocast(dtype=torch.float16):
                    value_loss_td, sampled_tensordict = value_loss(sampled_tensordict)
                scaler3.scale(value_loss_td["loss_value"]).backward()
                scaler3.unscale_(value_opt)
                clip_grad_norm_(value_model.parameters(), cfg.grad_clip)
                scaler3.step(value_opt)
                value_opt.zero_grad()
                scaler3.update()
```